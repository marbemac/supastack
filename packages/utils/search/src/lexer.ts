import type { Rules, Token } from 'moo';
import moo from 'moo';

export interface LexerToken<T extends LexerTypes = LexerTypes> extends Token {
  type: T;
}

export interface SimpleLexerToken<T extends LexerTypes = LexerTypes> extends Pick<Token, 'value'> {}

type StatesT = typeof STATES;
export type LexerTypes =
  | keyof StatesT['main']
  | keyof StatesT['qualifier']
  | keyof StatesT['literalQualifier']
  | keyof StatesT['multiValQualifier'];

export type LexerQualOps = (typeof QUAL_OPS)[number];

const qualVal = { match: /[a-zA-Z0-9@_\-\\.]+/, value: (s: string) => s.trimStart() };
const qualValWithSpace = { match: /[a-zA-Z0-9@_\s\\.]+/, value: (s: string) => s.trimStart(), lineBreaks: true };

const QUAL_OPS = ['>', '<', '>=', '<='] as const;
const qualOp = { match: [...QUAL_OPS] };

const STATES = {
  main: {
    qualKey: {
      match: /[a-zA-Z0-9\-_\\.]+:/,
      push: 'qualifier',
      // remove the : from the end
      value: s => s.slice(0, -1),
    },
    string: { match: /[a-zA-Z0-9_\\.]+/ },
    space: { match: /\s+/, lineBreaks: true },
  } satisfies Rules,

  qualifier: {
    quote: { match: '"', next: 'literalQualifier' },
    lbracket: { match: '[', next: 'multiValQualifier' },
    qualOp: { ...qualOp },
    qualVal: { ...qualVal, pop: 1 },
  } satisfies Rules,

  literalQualifier: {
    qualOp: { ...qualOp },
    qualVal: { ...qualValWithSpace, lineBreaks: true },
    quote: { match: '"', pop: 1 },
  } satisfies Rules,

  multiValQualifier: {
    qualOp: { ...qualOp },
    qualVal: { ...qualValWithSpace, lineBreaks: true },
    comma: ',',
    rbracket: { match: ']', pop: 1 },
  } satisfies Rules,
};

export const createLexer = () => {
  return moo.states(STATES);
};
