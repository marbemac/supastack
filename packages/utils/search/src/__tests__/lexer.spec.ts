import { expect, it } from 'vitest';

import { createLexer } from '../lexer.ts';
import { cases } from './cases.ts';

const prettyPrintLexer = (text: string) => {
  const lexer = createLexer();

  lexer.reset(text);

  const pretty = [];
  for (const r of lexer) {
    if (['qualKey', 'qualVal', 'string', 'qualOp'].includes(r.type || '')) {
      pretty.push(`[${r.type}] ${r.value}${r.value !== r.text ? ` (text=${r.text})` : ''}`);
      continue;
    }

    pretty.push(`[${r.type}]`);
  }

  return pretty;
};

it.each(cases)('%s', (_, input) => {
  expect({
    input,
    result: prettyPrintLexer(input),
  }).toMatchSnapshot();
});
