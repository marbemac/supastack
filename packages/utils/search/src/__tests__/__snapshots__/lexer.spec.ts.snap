// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`supports .-_ in qualifiers 1`] = `
{
  "input": "payments.plan-type_is:enterprise",
  "result": [
    "[qualKey] payments.plan-type_is (text=payments.plan-type_is:)",
    "[qualVal] enterprise",
  ],
}
`;

exports[`supports @.-_ in qualifier values 1`] = `
{
  "input": "email:jane@example.com type:public-company_x",
  "result": [
    "[qualKey] email (text=email:)",
    "[qualVal] jane@example.com",
    "[space]",
    "[qualKey] type (text=type:)",
    "[qualVal] public-company_x",
  ],
}
`;

exports[`supports freeform search mixed throughout 1`] = `
{
  "input": "freeform1 age:5 freeform2 search",
  "result": [
    "[string] freeform1",
    "[space]",
    "[qualKey] age (text=age:)",
    "[qualVal] 5",
    "[space]",
    "[string] freeform2",
    "[space]",
    "[string] search",
  ],
}
`;

exports[`supports multiple values on the same key 1`] = `
{
  "input": "user:[jane, john doe] release:[12.0]",
  "result": [
    "[qualKey] user (text=user:)",
    "[lbracket]",
    "[qualVal] jane",
    "[comma]",
    "[qualVal] john doe (text= john doe)",
    "[rbracket]",
    "[space]",
    "[qualKey] release (text=release:)",
    "[lbracket]",
    "[qualVal] 12.0",
    "[rbracket]",
  ],
}
`;

exports[`supports operators 1`] = `
{
  "input": "num_members:>5.0",
  "result": [
    "[qualKey] num_members (text=num_members:)",
    "[qualOp] >",
    "[qualVal] 5.0",
  ],
}
`;

exports[`supports qualifiers 1`] = `
{
  "input": "plan:enterprise age:5",
  "result": [
    "[qualKey] plan (text=plan:)",
    "[qualVal] enterprise",
    "[space]",
    "[qualKey] age (text=age:)",
    "[qualVal] 5",
  ],
}
`;

exports[`supports quoted qualifiers 1`] = `
{
  "input": "user:\\"john doe\\"",
  "result": [
    "[qualKey] user (text=user:)",
    "[quote]",
    "[qualVal] john doe",
    "[quote]",
  ],
}
`;

exports[`treats freeform as a string 1`] = `
{
  "input": "freeform",
  "result": [
    "[string] freeform",
  ],
}
`;
